[BEAM-425] Add ElasticsearchIO

echauchot opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

R: @jbonofre

jbonofre commented:
CC: @jkff 
CC: @dhalperi

jkff commented:
I briefly skimmed the code and this looks awesome - everything major that there is to do right, is done right! I'll post a number of minor comments later today.

jbonofre commented:
@jkff thanks for the first pick ;) Ready to address your comment ;)

jkff commented on sdks/java/io/elasticsearch/contrib/ElasticsearchIOIT.java:
```diff
@@ -0,0 +1,106 @@
+            ElasticsearchIO.ConnectionConfiguration
+                .create("http://" + ES_IP + ":" + ES_HTTP_PORT, ES_INDEX, ES_TYPE))
+            .withScrollKeepalive("5m"));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(96253L);
```
Explain where the magic number comes from?

jbonofre replied:
It's the number of documents loaded by logstash (part of the integration tests). I will add a comment to explain this.

echauchot replied:
Actually, integration tests are a first shot (not finished yet) mainly there to discuss (as part of Stephen mails around integration tests). Among other things, I will make logstash ingestion stop at certain suitable number of documents inserted and update the asserts accordingly

jkff commented on sdks/java/io/elasticsearch/contrib/ElasticsearchIOIT.java:
```diff
@@ -0,0 +1,106 @@
+    ElasticsearchIO.Read read =
+        ElasticsearchIO.read()
+            .withConnectionConfiguration(
+                ElasticsearchIO.ConnectionConfiguration
```
You may want to extract the ConnectionConfiguration into a class-level constant.

echauchot replied:
ok, I'll do so

jkff commented on sdks/java/io/elasticsearch/contrib/ElasticsearchIOIT.java:
```diff
@@ -0,0 +1,106 @@
+              .create("http://" + ES_IP + ":" + ES_HTTP_PORT, ES_INDEX, ES_TYPE));
+    ElasticsearchIO.BoundedElasticsearchSource initialSource =
+        new ElasticsearchIO.BoundedElasticsearchSource(read, null);
+    assertTrue("Wrong estimated size", initialSource.getEstimatedSizeBytes
```
assertThat(..., gt(1900000))

Also explain 1) why the size should be greater than that and 2) why that can be relied on so that the test is stable (does ElasticSearch guarantee that these statistics are always up-to-date?)

jbonofre replied:
Elasticsearch never uses the same size even using the same index names, the same document, the same number of documents. So, we are using a lower bound in the itests.

echauchot replied:
I added a comment on the integration test

jbonofre replied:
@echauchot no need to comment when you did the change: the comments around the line will be closed when you will push your changes.

echauchot replied:
Yes sure :)

ssisk replied:
I would like to advocate for testing as few things as possible in the integration tests - do we think this test of the value of getEsimatedSizeBytes is adding additional important checks above and beyond the test on getEstimatedSizeByte that we have in the Test.java?

jkff commented on sdks/java/io/elasticsearch/contrib/create_elk_container.sh:
```diff
@@ -0,0 +1,26 @@
+#    limitations under the License.
+#
+################################################################################
+docker create -p 5601:5601 -p 9200:9200 -p 5044:5044 -p 5000:5000 -p 9300:9300 -it -v $(pwd):/home/$USER/ --name elk-2.4  sebp/elk:es240_l240_k460
```
Some comments in this file wouldn't hurt too. If it's just copied over from somewhere, link to the source and explain the modifications.

jbonofre replied:
Yes, I will document that a bit.

jkff commented on sdks/java/io/elasticsearch/contrib/logstash.conf:
```diff
@@ -0,0 +1,53 @@
+#    limitations under the License.
+#
+################################################################################
+input {
```
Ditto

echauchot replied:
ok, I'll document here as well

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+    }
+
+    public static ConnectionConfiguration create(String[] addresses, String index, String type) {
+      checkArgument(addresses != null, "ConnectionConfiguration.create(addresses, index, type) "
```
No need to repeat in the message what a failure stack trace will already tell you.
Simply:
`checkNotNull(addresses, "addresses");`
`checkArgument(addresses.length != 0, "addresses must not be empty")`
`checkNotNull(index, "index")`
etc.

jbonofre replied:
We discussed about the messages with Ken and we agreed about that `checkNotNull` are not useful for end users.
See BEAM-959.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+    abstract List<String> getAddresses();
+    @Nullable abstract String getUsername();
+    @Nullable abstract String getPassword();
+    abstract String getIndex();
```
Just to confirm: all of these fields make sense both for read() and write(), correct?

jbonofre replied:
Correct.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+ * }</pre>
+ *
+ * <p>The connection configuration also accepts optional configuration: {@code withUsername()} and
+ * {@code withPassword()}.
```
How does authentication work if you don't supply username and password?

jbonofre replied:
Elasticsearch doesn't use authentication by default. So, `username` and `password` are not required in that case.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+    }
+
+    public Read withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "ElasticsearchIO.read()"
```
Same comments for all checkBlah() calls in this file.

jbonofre replied:
See BEAM-959.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+
+    private ElasticsearchIO.Read spec;
+    @Nullable
+    private final String shardPreference;
```
Document this? (the word "preference" makes me worry - does ES guarantee that reading with a different "shardPreference" is guaranteed to give non-overlapping results?)

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+   */
+  protected static class BoundedElasticsearchSource extends BoundedSource<String> {
+
+    private ElasticsearchIO.Read spec;
```
final

jbonofre replied:
Indeed, I missed this one ;)

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+
+      // we split per shard
+      // unfortunately, Elasticsearch 2.x doesn't provide a way to do parallel reads on a single
+      // shard
```
Explain a bit more about splitting here - e.g. what are "primary" shards? Perhaps link to the official docs about this?

jbonofre replied:
We experimented different approach for the splitting.
Basically, Elasticsearch stores the index on a shards.
To optimize the read, we create a source per shard. At the beginning we wanted to create several source per shards, depending of the shard size, (each source reading a subset of documents), however, unfortunately, it's not possible to "split" inside a shard.
So basically, the split is one source per shard.
The number of shards is defined by Elasticsearch (by index) depending of the routing.

echauchot replied:
For primary shards: let's say the ES index is configured to split data into 5 shards with a replica of 1 (for failover). In that case, there will be a total of 10 shards (5 primary, 5 replica). Each beam source reads the documents contained in one shard. As we don't want to have duplicate documents in PCollection, we only read primary shards. In case of failover (ES side), primary shards stored in a failing ES node are no longer available and replica shards (stored in other nodes) become primary nodes.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+      // as Elasticsearch 2.x doesn't not support any way to do parallel read inside a shard
+      // the estimated size bytes is not really used in the split into bundles.
+      // However, we implement this method anyway as the runners can use it.
+      // NB: Elasticsearch 5.x now provides the slice API.
```
What is the slice API?

jbonofre replied:
The slide API is a new API coming into Elasticsearch 5. It will allow to not use shard anymore and use a Elasticsearch API really focusing on split read.

echauchot replied:
On this version of the IO (for ES 2.4), we do not use desiredBundleSize because we cannot split shards. With the slice API in ES 5.0 we will be able to use desiredBundleSize. Basically we will just ask the slice API to return data in nbBundles = estimatedSize / desiredBundleSize chuncks. So each beam source will read around desiredBundleSize volume of data.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+      // NB: Elasticsearch 5.x now provides the slice API.
+      JsonObject statsJson = getStats(false);
+      JsonObject indexStats =
+          statsJson.getAsJsonObject("indices").getAsJsonObject(
```
Please link to docs here too, e.g. to docs explaining format of the response you're parsing here.

jbonofre replied:
Agree, it could be helpful for the reviewer and future contributors.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+        params.put("level", "shards");
+      }
+      String endpoint = String.format("/%s/_stats",
+                                      spec.getConnectionConfiguration
```
Weird formatting

jbonofre replied:
You mean in term of code style, right ?

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+                                      source.spec.getConnectionConfiguration().getIndex(),
+                                      source.spec.getConnectionConfiguration().getType());
+      Map<String, String> params = new HashMap<>();
+      params.put("scroll", source.spec.getScrollKeepalive() != null ? source.spec
```
Hmm, what does this keepalive thing mean? Does it mean that if you take more than 1m between calls to start()/advance(), the read will fail? A runner can definitely take more than that between such calls - e.g. if you're reading and applying a long DoFn to the results and if a runner executes that as "start, process, advance, process, advance, process, ...".

echauchot replied:
Good point. 
When using scroll API, ES needs a statefull scroll context server side. This keepalive is the delay after which it will remove the context. This delay needs to be long enough to process the chunk of data returned by previous scroll request. In our case - for start() and advance() - each chunk contains only 1 doc. So if the processing of a single doc is longer than 1 minute before next advance() - not next start() because each start() starts a new scroll - , the read will fail indeed. That is why we allow the user to specify a keepalive in case his processing of one doc is longer than default 1 minute.

jbonofre replied:
I don't see a problem there: if the user design a pipeline with a "long" processing, then he can extend the scroll keepalive. The scroll is "keepalive" is required in the `advance()`.

jbonofre replied:
By the way, each use of the scroll extends the keepalive for a new round.

echauchot replied:
Yes I don't see any problem either

jkff replied:
This is unfortunate because it means the only proper way to set this value for the user is trial-and-error: Beam doesn't give any guarantees on how long it will take between successive advance() calls. Even if you assume that it's "as soon as possible", the user still has no way to predict a correct value in advance, because e.g. in case of any runner that supports fusion, the answer will be "however long it will take to execute the composition of DoFn's the runner decided to fuse on top of this read" and this can not be predicted by the user. It also goes against the "no knobs" spirit of Beam - things should "just work" without tweaking parameters like batch sizes, keepalive timeouts, etc.

However, I don't see a good way to get rid of this particular parameter (the best I can come up with is "materialize the PCollection so that there's no fusion going on and Beam reads documents as fast as possible, and cross fingers", but this has its downsides).

Just please document that users should change it ONLY if their pipeline is failing with this particular error (...quote the error that happens if keepalive expires...), and never for any other reason.

Same comment applies to batch sizes. Users shouldn't be tweaking them unless the pipeline behaves unacceptably for a reason that can only be fixed by changing batch sizes.

echauchot replied:
Thanks for the example of the fusion, I was only thinking about volume of messages to process (1 message in ES IO case).
+1: I will document it and increase the default scroll timeout to 5 minute to reduce the probability of timeout.
+1 on batchSizes as well

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+    @Override
+    public boolean advance() throws IOException {
+      String requestBody = String.format("{\"scroll\" : \"%s\",\"scroll_id\" : \"%s\"}",
+                                         source.spec.getScrollKeepalive() != null ? source.spec
```
Hm, I don't quite like that this snippet (defaulting to 1m) is repeating. Can you structure the code in a way that getScrollKeepalive() is simply "1m" by default?

jbonofre replied:
Yes, we can set this in the spec.

jbonofre replied:
I would do `build().setKeepAlive("1m").build()` when creating.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+    private boolean updateCurrent(JsonObject searchResult) {
+      //stop if no more data
+      if (searchResult.getAsJsonObject("hits").getAsJsonArray("hits").size() == 0) {
+        return false;
```
current = null;

jbonofre replied:
Missed this, thanks !

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+
+    @Override
+    public String getCurrent() throws NoSuchElementException {
+      return current;
```
if (current == null) throw new NoSuchElementException();

jbonofre replied:
Yup, missed this one too.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+        // error is ES 5.x so protect this: cannot request if a particular scroll id is still open
+        // so ignores exception (raised with ES v5.x)
+        if (e.getMessage() != null && !e.getMessage().contains("no scroll ids specified")) {
+          throw e;
```
In this case will you fail to close restClient?

echauchot replied:
yes, thanks! restclient.close() should be in a finally bloc

jbonofre replied:
@echauchot I already fixed that. Gonna push.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+                                  Collections.<String, String>emptyMap(),
+                                  entity, new BasicHeader("", ""));
+      } catch (IOException e) {
+        //ES gives the same scroll id to each source, close that scroll more than once produces an
```
I don't quite understand this comment. Same scroll id across shards, or across something else?
Is there a piece of docs you can link to here?

echauchot replied:
Well, I did not find any docs about that, this is just a behavior that I observed, let me rephrase: each reader.start() starts a new scroll identified by a scroll_id (given by ES). What I observed is that ES gives the same scroll_id to multiple readers. As a consequence, when these readers call close() they might delete a scroll that has already been deleted by another reader (that received the same scroll_id). This does no cause any exception to be thrown with ES 2.x but with ES 5.x, deleting an already deleted scroll throws an Exception. We ignore this exception if it is thrown

jkff replied:
Actually, getting the same scroll_id from multiple readers sounds really really bad. Doesn't it mean that we're also getting data corruption (duplicate data) from these readers? I looked at ES open issues and didn't find something like this; I'd be surprised if this is happening for everybody but nobody except us noticed it here.

I would strongly prefer that this problem be debugged before submitting this PR - either isolated to a publicly linked bug in ES and using a workaround here in code, or if it's just a bug in this connector, then fixed.

jkff replied:
(in another thread you say it turned out to indeed be a bug in the IO)

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Write withBatchSize(long batchSize) {
```
Document this and the other batch size method - specifically, what should a user know about setting them?

jkff replied:
Hmm, I'm not sure where you added this - the methods withBatchSize and withBatchSizeMegabytes still don't have any javadoc.

echauchot replied:
I added doc in the class javadoc not on the method. But I will add javadoc for every with* end user method

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+
+    private static class WriterFn extends DoFn<String, Void> {
+
+      private Write spec;
```
final

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+      @FinishBundle
+      public void finishBundle(Context context) throws Exception {
+        StringBuilder bulkRequest = new StringBuilder();
+        if (batch.size() > 0) {
```
Invert this to reduce nesting?

jkff replied:
I mean transform code into `if (batch.isEmpty()) return;` and reduce nesting of lines 587-621, so the code will be easier to read - e.g. the current code kept me holding my breath for "is there an else branch, or anything after the if?"

echauchot replied:
ok, of course, better that way indeed

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+          JsonObject searchResult = parseResponse(response);
+          boolean errors = searchResult.getAsJsonPrimitive("errors").getAsBoolean();
+          if (errors) {
+            throw new IllegalStateException("Can't update Elasticsearch");
```
Is it possible to include the actual error information in this message?

jbonofre replied:
Yes, we can get error message from JSON.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+          if (errors) {
+            throw new IllegalStateException("Can't update Elasticsearch");
+          }
+          batch.clear();
```
currentBatchSizeBytes = 0?

echauchot replied:
yes, thanks, forgot that one

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,595 @@
+
+      @Setup
+      public void createClient() throws Exception {
+        if (restClient == null) {
```
It can't be non-null here, so the if() is unnecessary.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,302 @@
+            .put("index.store.stats_refresh_interval", 0);
+    node = NodeBuilder.nodeBuilder().settings(settingsBuilder).build();
+    LOGGER.info("Elasticsearch node created");
+    if (node != null) {
```
What does it mean if it's null? Is silently-do-nothing the proper action if it's null?

echauchot replied:
yes, thanks again, it is better to raise an error because if node creation failled, all the others tests will fail.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,302 @@
+    }
+    final BulkResponse bulkResponse = bulkRequestBuilder.execute().actionGet();
+    if (bulkResponse.hasFailures()) {
+      throw new IOException("Cannot insert samples in index " + ES_INDEX);
```
Include failure info in the message too

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,302 @@
+    }
+  }
+
+  private void sampleIndex(long nbDocs) throws Exception {
```
The name is ambiguous - is "sample" a noun or a verb? Perhaps "prepareIndex" would be a better name.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,302 @@
+    PipelineOptions options = PipelineOptionsFactory.create();
+    String[] addresses = {"http://" + ES_IP + ":" + esHttpPort};
+    ElasticsearchIO.Read read =
+        ElasticsearchIO.read().withConnectionConfiguration(
```
Extract this into a class-level variable

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,302 @@
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    sampleIndex(NB_DOCS);
```
This is repeating - should it be in `@Before`?

echauchot replied:
No because testWrite* tests do not need to insert a dataset because their point is to insert data.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,302 @@
+    Pipeline pipeline = TestPipeline.create();
+
+    String[] addresses = {"http://" + ES_IP + ":" + esHttpPort};
+    String[] scientists =
```
Extract this (creating the `data` list) into a helper method to dedup it against `sampleIndex`?
E.g. create a method `List<String> createDocuments(int numDocs)` and a separate method `createIndexWithDocuments(List<String> docs)`

echauchot replied:
yes good suggestion, I forgot to deduplicate this part of test code. Thanks!

echauchot commented:
Thanks @jkff and @jbonofre for the review and all the comments! I'll make the advised modifications.

jkff commented:
OK, please let me know when tests are green and when the PR is ready for another round of review.

echauchot commented:
Well, tests of elasticsearch IO are already green. Jenkins failure above is about JDBC IO which is not part of that pull request. Strange, I will take a look.  
Besides, I have not finished updating the code with some of your advices yet. I'll keep you posted

echauchot commented:
Hi @jkff, the PR is ready for another round of review. Basically I addressed all your comments and questions except the one on nesting. Can you please take a look at my comment above ? Thanks

jbonofre commented:
Gonna take a look as well.

echauchot commented:
R: @jbonofre

ssisk commented on sdks/java/io/elasticsearch/contrib/ElasticsearchIOIT.java:
```diff
@@ -0,0 +1,106 @@
+  @Category(NeedsRunner.class)
+  public void testReadVolume() throws Exception {
+    //use flink runner otherwise only one source is created
+    String[] args = new String[] { "--runner=FlinkRunner", "--project=test-project" };
```
directrunner changed recently so that it should try to split sources - have you tried direct runner recently? (I'm betting you did, but thought I'd double check) BEAM-833 should have addressed this. If not, we can follow up with thomas groh about why that isn't working.

ssisk replied:
I think there are better ways to get a particular test run on a particular set of runners. 

WordCountIT is an example of this: https://github.com/apache/incubator-beam/blob/master/examples/java/pom.xml#L144 

It is able to specify how the test is run. Unfortunately we don't have that well documented yet - I'm adding to the documentation we want to add on the website as we work on this part of things. I took a brief note of this here - https://docs.google.com/document/d/153J9jPQhMCNi_eBzJfhAg-NprQ7vbf1jNVRgdqeEE8I/edit#heading=h.pv4fkiv0m0tl

jbonofre replied:
IMHO, the IT should not be directly in the IO: we should have a ITs module covering the IT.
It's what I'm doing in Apache Karaf: each module has utest, all ITs (for most/all modules) are in itests module, dedicated to IT.

echauchot replied:
@ssisk: thanks for your remarks! I know passing command line arguments was hacky, I did not know any cleaner way to do it. But, indeed, configuring failsafe maven plugin to pass pipeline options is a lot cleaner. Thanks for the pointer.  

For the use of flink runner in place of direct runner: I have just re-tested with direct runner and, direct runner now calls splitIntoBundles() and thus creates the expected number of sources/readers so I can use it in place of flink runner.

echauchot replied:
@jbonofre I am kind of indecisive about where to store IT material. What you suggest might be better. Maybe we should discuss this in Stephen thread about IT in the ML?

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,635 @@
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
```
Do ES documents have a timestamp? perhaps makes sense to do getCurrentTimestamp() too.

echauchot replied:
No by default es docs do not have a timestamp. You mean calling getCurrentTimestamp() and let it throw the exception?

jkff replied:
No, if you can't provide a meaningful timestamp then just don't override the function.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,635 @@
+                                  entity, new BasicHeader("", ""));
+      } catch (IOException e) {
+          // Each reader.start() starts a new scroll identified by a scroll_id .ES sometimes gives
+          // the same scroll_id to multiple readers. As a consequence, when these readers call
```
This is weird - does it mean closing one reader may make another reader stop working, because its scroll has been deleted and its search context is not being kept alive anymore? That sounds like a very serious bug in Elasticsearch if it is so; is there a public bug entry about it somewhere?

echauchot replied:
Yes weird indeed, but this is what I observed. I asked the question to a member of Elasticsearch dev team. Waiting for reply.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,635 @@
+        StringBuilder bulkRequest = new StringBuilder();
+        if (batch.size() > 0) {
+          for (String json : batch) {
+            bulkRequest.append(json);
```
I suppose the only reason we need to control batch sizes is so that this bulkRequest doesn't get too large, right?

In that case, perhaps, we can remove batch size parameters and just flush when bulkRequest gets too large instead?

Per https://www.elastic.co/guide/en/elasticsearch/guide/current/indexing-performance.html#_using_and_sizing_bulk_requests seems like we can just use 10MB, period.

echauchot replied:
Actually, if a batch contains exactly batchSize documents or batchSizeMagabytes bytes, finishBundle() gets called one more time after that batch with an empty batch.  insert with empty batch in ES do not work. I protected with if() to avoid bulk request when batch is empty. 

For the sizes removal: well, I agree batchSize in number of docs is not so relevant for performance (because doc size changes), but batchSizeInMegabytes is. As explained in the above doc, end user might need to configure it to get best performance. And I remember that, as a user, I had to do this in a project to improve ES indexation time. The fact is that both are present is ES bulk API. [https://www.elastic.co/guide/en/elasticsearch/client/java-api/2.4/java-docs-bulk-processor.html](url).See setBulkActions and setBulkSize.

Besides, in the IO I set defaults to 1000 docs and 5 MB

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,635 @@
+          JsonObject searchResult = parseResponse(response);
+          boolean errors = searchResult.getAsJsonPrimitive("errors").getAsBoolean();
+          if (errors) {
+              StringBuilder errorMessages = new StringBuilder("Error writing to Elasticsearch, "
```
Indentation here is off, too. And in a bunch more places. And a bunch of whitespace is also off (e.g. line 607).

I usually run my code through [google-java-format](https://github.com/google/google-java-format) to stop worrying about these issues. It's not specifically a mandated code style, but at least it gives a reasonable formatting, consistency and convenience (it has IDE plugins too).

echauchot replied:
ok, thanks for pointing me at it, it might better than intelliJ standard formatter.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,635 @@
+                    errorMessages.append(String.format("%n document id %s : ", docId));
+                      String errorMessage = error.getAsJsonObject("caused_by")
+                                            .getAsJsonPrimitive("reason").getAsString();
+                      errorMessages.append(errorMessage);
```
Are we producing a raw concatenation of errors without any delimiter?

echauchot replied:
No, actually the delimiter is \n. The output of testWriteWithErrors() for 2 malformed documents is 

Error writing to Elasticsearch, some elements could not be inserted:
 document id AVjASLucVlDMC7b3bq3B : Unexpected character (';' (code 59)): was expecting a colon to separate field name and value
 at [Source: org.elasticsearch.common.io.stream.InputStreamStreamInput@714b6929; line: 1, column: 14]
 document id AVjASLucVlDMC7b3bq3E : Unexpected character (';' (code 59)): was expecting a colon to separate field name and value
 at [Source: org.elasticsearch.common.io.stream.InputStreamStreamInput@3bce5cae; line: 1, column: 14]

The only thing I can get in addition to the error message is the doc id. I cannot get the doc _source because it is not in ES error response. If the user does not provide a doc id (if ES generates one at insertion time) that might not help the user identify failing doc. But this is the best information we can give.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+/**
+ * Test on {@link ElasticsearchIO}.
+ */
+public class ElasticsearchIOTest implements Serializable {
```
`@RunWith(JUnit4.class)`

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(ElasticsearchIOTest.class);
+
+  private static final String DATA_DIRECTORY = "target/elasticsearch";
+  public static final String ES_INDEX = "beam";
```
Why are some of these private and some public? They should all be private.

echauchot replied:
Sure, it's an error.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+    ArrayList<String> data = new ArrayList<>();
+    for (int i = 0; i < nbDocs; i++) {
+      int index = i % scientists.length;
+      //insert 2 malformed documents is asked to do so
```
Don't understand this comment

echauchot replied:
bad copy/paste, thanks for noting

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+    Client client = node.client();
+    final BulkRequestBuilder bulkRequestBuilder = client.prepareBulk().setRefresh(true);
+    List<String> data = createDocuments(nbDocs, false);
+    for (String document:data){
```
Whitespace in this line and in general in this file is also inconsistent - please reformat it as well.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+  private static ElasticsearchIO.ConnectionConfiguration connectionConfiguration;
+
+  @BeforeClass
+  public static void beforeClass() throws IOException {
```
What's the main difference between this one and `IT`? The code looks really similar, I don't quite like the duplication. What is the `IT` testing that this one isn't?

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(connectionConfiguration)
+                              .withScrollKeepalive("5m"));
```
Why is it necessary to set the keepalive in this test - would it fail if you hadn't set it?

echauchot replied:
No of course, it would not fail with default 1m, because 1 mintute is long enough to process 1 document (see previous comment). This line is added just to test scrollKeepalive option setting. I'll add a comment

echauchot replied:
No, the test would not fail because default 1m is enough to process 1 document. So you're right withScrollKeepalive("5m") is not critical for this test, I'll remove it if it is misleading about the test's intention.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(connectionConfiguration)
+                              .withQuery(query));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(NB_DOCS / 10);
```
Extract 10 into a constant (NUM_SCIENTISTS?)

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+      .withConnectionConfiguration(connectionConfiguration));
+    pipeline.run();
+
+    // upgrade
```
This comment doesn't explain what's going on. Elaborate?

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NB_DOCS, false);
```
I keep forgetting what the boolean parameter in this method stands for. It's better to replace it with an enum - then the code will read like `createDocuments(NUM_DOCS, DO_NOT_INJECT_INVALID_DOCS or INJECT_SOME_INVALID_DOCS)`

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+    // upgrade
+    node.client().admin().indices().upgrade(new UpgradeRequest()).actionGet();
+
+    SearchResponse response = node.client().prepareSearch().execute().actionGet(5000);
```
Do SearchResponse objects need to be closed or something?

echauchot replied:
No need, SearchResponse is just a wrapper object for the json response.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+      withConnectionConfiguration(connectionConfiguration));
+    try {
+      pipeline.run();
+      fail("Exception should have been thrown");
```
Use the `ExpectedException` rule instead of manual try/catch.

echauchot replied:
Yes! That's what I was looking for instead of try catch. Thanks.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+        //type of e is Pipeline$PipelineExecutionException
+
+      String message = e.getCause().getMessage();
+      assertTrue("Wrong exception thrown", e.getCause().getClass() == IOException.class);
```
Use most specific mockito matchers (such as `contains`) instead of simple assertTrue/assertFalse to get much better error messages.

echauchot replied:
ok, a standard hamcrest matcher for Exception class and a custom hamcrest matcher for regExp should to it better.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,303 @@
+    PDone collection = pipeline.apply(Create.of(data)).apply(
+        ElasticsearchIO.write()
+            .withConnectionConfiguration(connectionConfiguration)
+            .withBatchSize(NB_DOCS / 2)
```
Same comment here - do not set parameters that are not critical for the test (i.e. if you don't set them, the test won't run at all, or won't test what you want it to test).

I don't quite understand where `NB_DOCS/2` is coming from: the runner may split `data` into more than 1 bundle, so this code currently isn't accomplishing the goal of forcing there to be more than 1 batch per bundle.

echauchot replied:
-  +1 for non critical parameters (like for previous comment)

- Considering withBatchSize*() tests: I thought I could force the number of bundles but I realize that this is runner specific, so the code above is pointless. In fact withBatchSize*() are more with**Max**BatchSize() because the runner may split data before specified batchSize is reached. So it might be misleading for users to have these parameters in the API. They might expect that they behave the same as corresponding ElasticSearch bulk parameters (create fixed size batches). @jkff, @jbonofre any thoughts? Should they be kept? If yes, how should they be properly tested?

jkff replied:
Perhaps unit-test the fn using DoFnTester? And yes, renaming to withMaxBatchSize and documenting that it's only a hint, sounds like a good idea.

jbonofre replied:
I always consider as max batch size but agree to explicitly rename to withMaxBatchSize().

jbonofre replied:
For the test, as BundleFactory is not available outside of the runners, I would use the DoFnTester util.

echauchot replied:
Yes that's what I was looking for to test the writer :) I'll use DoFnTester in conjunction with mockito to count how many times finishBbundle() was called ( = number of write bundles).

echauchot commented:
@jkff , @ssisk , thank you guys for the comments! I'll address them and keep you posted

echauchot commented on sdks/java/io/elasticsearch/contrib/logstash.conf:
```diff
@@ -0,0 +1,61 @@
+    ]
+    # count is multiplied by threads (!)
+    # By default ES will create 5 shards, so beam IO will create 5 sources.
+    # Inject 60 000 docs so that each source reads more docs than ES read window (10 000)
```
Well, we had problems on high volume (hence integration tests) before implementing the IO with scroll. Indeed with standard from/size reading, ES uses a window of 10 000 documents by default. So from and size parameters cannot be more than 10 000. That is why we chose to use scroll instead. Scroll do not use ES window. So there should not be any problem about that. Nevertheless I wanted to test with a lot of documents for each source so that we detect any problem with high volume. I will remove the comment about window (because of the use of scroll) and only explain about the high volume.

echauchot commented on sdks/java/io/elasticsearch/contrib/logstash.conf:
```diff
@@ -0,0 +1,61 @@
+
+filter {
+  ruby {
+    code => "event.set('gen_id', rand(100000 - 1) + 1)"
```
I realized that this part of code is unnecessary because ES will generate unique ids at insertion time. So I just removed it, thanks for pointing me at it.

echauchot commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,635 @@
+  /**
+   * A {@link BoundedSource} reading from Elasticsearch.
+   */
+  protected static class BoundedElasticsearchSource extends BoundedSource<String> {
```
Yes! thanks, I did not know that VisibleForTesting

echauchot commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,635 @@
+       Each beam source reads the documents contained in one shard.
+       As we don't want to have duplicate documents in PCollection, we only read primary shards.
+       In case of failover (ES side), primary shards stored in a failing ES node
+       are no longer available and replica shards (stored in other nodes) become primary shards.
```
I just talked to one guy of elasticsearch dev team. In fact, my comment is partially wrong: primary and replica shards have the same shard_id. So the even if we specify preference=shards:2, ES load balances (round robin) the request between primary shard 2 and replica shard 2. But, as each shard is responsible for only one part of the data, there will be no duplicate.

jbonofre replied:
Yes, it's what we discussed and noticed.

echauchot commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,635 @@
+ * <h3>Reading from Elasticsearch</h3>
+ *
+ * <p>ElasticsearchIO source returns a bounded collection of String representing JSON document
+ * as {@code PCollection<String>}.
```
No, actually, Elasticsearch API (elasticsearch rest client in fact) returns a org.apache.http.HttpEntity see response.getEntity() in method ElasticsearchIO.parseResponse(). We parse it using Gson.

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,744 @@
+        if (errors) {
+          StringBuilder errorMessages =
+              new StringBuilder(
+                  "Error writing to Elasticsearch, " + "some elements could not be inserted:");
```
This string literal too. See if you can find others (e.g. Ctrl+F, `" + "`)

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,336 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ElasticsearchIOTest.class);
+
+  private static final String DATA_DIRECTORY = "target/elasticsearch";
```
It's better to use a proper temp directory for this, e.g. use the `@Rule TemporaryFolder`.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,336 @@
+  }
+
+  @Test
+  @Category(RunnableOnService.class)
```
Same note for these tests - make sure they really run.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,336 @@
+    fnTester.processBundle(input);
+    int numWriteBundles = (int) (NUM_DOCS / BATCH_SIZE);
+    // +1 because processBundle calls finishbundle()
+    //TODO do not pass because of TestContext extends OldDoFn.Context whereas Context is in DoFn
```
Please rebase onto https://github.com/apache/incubator-beam/pull/1461

echauchot replied:
Actually, my comment appeared to be incorrect. The problem was the inner cloning of the doFn by DoFnTester. As a consequence mockito did not see any interaction with the spy. I've fixed that using  fnTester.setCloningBehavior(DoFnTester.CloningBehavior.DO_NOT_CLONE);
I'll push this change with all the ongoing commits.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,336 @@
+    int numWriteBundles = (int) (NUM_DOCS / BATCH_SIZE);
+    // +1 because processBundle calls finishbundle()
+    //TODO do not pass because of TestContext extends OldDoFn.Context whereas Context is in DoFn
+    verify(writerFn, times(numWriteBundles + 1)).finishBundle(Matchers.any(DoFn.Context.class));
```
I'd rather recommend to feed it elements one-by-one and verify that every time after you feed BATCH_SIZE elements, the elements get inserted and appear in ES index. That would verify the behavior rather than implementation (black box vs. white box).

echauchot replied:
Actually, my comment appeared to be incorrect. The problem was the inner cloning of the doFn by DoFnTester. As a consequence mockito did not see any interaction with the spy. I've fixed that using  fnTester.setCloningBehavior(DoFnTester.CloningBehavior.DO_NOT_CLONE);
I'll push this change with all the ongoing commits.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,336 @@
+  }
+
+  @AfterClass
+  public static void afterClass() {
```
May be easier to read if you place this method next to the BeforeClass method.

echauchot commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,744 @@
+              source.spec.getConnectionConfiguration().getType());
+      Map<String, String> params = new HashMap<>();
+      params.put("scroll", source.spec.getScrollKeepalive());
+      params.put("size", "1");
```
Yes, that's funny you're noticing that :) I was just in the middle of coding a buffer to store read docs and make the scroll read 1000 docs at each request. I'll commit the code today.

echauchot commented on sdks/java/io/elasticsearch/contrib/create_elk_container.sh:
```diff
@@ -0,0 +1,37 @@
+#Create an ELK (Elasticsearch Logstash Kibana) container for ES v2.4 and compatible Logstash and Kibana versions,
+#bind then on host ports, allow shell access to container and mount current directory on /home/$USER inside the container
+
+docker create -p 5601:5601 -p 9200:9200 -p 5044:5044 -p 5000:5000 -p 9300:9300 -it -v $(pwd):/home/$USER/ --name elk-2.4  sebp/elk:es240_l240_k460
```
ports are the default ports, they can be busy of course. Need to check if there are available on jenkins servers, if not they can be changed in the above script. For the /etc/sysctl.conf part, it is mandatory for ES 5.0 to run but it can be configured outside the above script. I know nothing about our jenkins servers as I don't have access.

jkff replied:
I guess I don't understand the role of this script in this PR then. Is this script run as part of precommit integration tests on Jenkins? Should it be? Is it necessary for the `IT` test to run?

echauchot commented on sdks/java/io/elasticsearch/contrib/logstash.conf:
```diff
@@ -0,0 +1,58 @@
+      }'
+    ]
+    # count is multiplied by threads (!)
+    # By default ES will create 5 shards, so beam IO will create 5 sources.
```
By default, each index in Elasticsearch is allocated 5 primary shards. I'll add this comment in code

echauchot commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,744 @@
+
+  public static Write write() {
+    return new AutoValue_ElasticsearchIO_Write.Builder()
+        .setMaxBatchSize(1000L)
```
it is the elasticsearch default for bulk. I'll add this comment in code

echauchot commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,744 @@
+    public static ConnectionConfiguration create(String[] addresses, String index, String type) {
+      checkArgument(
+          addresses != null,
+          "ConnectionConfiguration.create(addresses, index, type) " + "called with null address");
```
Weird, I now use google autoformat. Previous autoformatter has split the strings and google auto format has put them in same line without merging them. I'll fix them all and check for google auto-format conf. Thanks for noting

echauchot commented on sdks/java/io/elasticsearch/contrib/ElasticsearchIOIT.java:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  @Test
+  @Category(RunnableOnService.class)
```
Done, yes they are run 
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0

jbonofre commented:
I will add a new feature to the `ElasticsearchIO`: add a way to define SSL/TLS configuration for the HTTP REST Client (for mutual authentication, ...).

echauchot commented:
Hi @jkff, I have addressed all your previous comments, I am ready for another round of review. I just need to figure out why the tests behave differently on others JDK (travis part). In the meantime can you take a look at the commits please? Thanks

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,788 @@
+     * @return the {@link Read} with batch size set
+     */
+    public Read withBatchSize(long batchSize) {
+      return builder().setBatchSize(
```
It's better to reject invalid values than silently change them to valid ones. Might as well also check that the value is `> 0`.

echauchot replied:
I have no reply button on previous comment, so I'll answer in this one:
Yes, actually, I put IT material in the PR to launch the discussion about integration tests. There is an ongoing discussion on the ML with JB and Stephen on where to put IT material (separate IT module or IO module). The script you mention is to be launched at backend instance creation time (it could be run only once or before running the IT test suite if we decide to recreate instances each time before running tests suites). As external ES instance is needed for the IT tests to run (and logstash ingestion also), then, yes, this script is needed for IT tests to run.

For above comment on withBatchSize: I aggree

ssisk replied:
eugene: if you have questions about this, let me know. 

Etienne - Since the IT is blocked on having a way to run the tests, would it make sense to pull the IT file + script into a separate PR? I don't want you to be blocked on the other discussion and it sounds like otherwise this PR is getting ready to submit.

echauchot replied:
@ssisk, Yes no pb it's fine for me to move them to another PR. Do I open a new Jira ticket for this?

ssisk replied:
It's probably easiest to track it with a new issue, so I created one - https://issues.apache.org/jira/browse/BEAM-1184 

if that's not helpful or you want to use the original issue (I'm fine either way), feel free to won't fix that one.

echauchot replied:
Perfect!

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,788 @@
+    @Override
+    public boolean advance() throws IOException {
+      if (batch.size() > 0){
+        current = batch.remove(0);
```
This is inefficient - removing the first element in an `ArrayList` is O(N). Just use an `Iterator` instead of `List` as the member variable.

echauchot replied:
Yes, sure, totally agree, thanks for pointing

jkff commented on sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:
```diff
@@ -0,0 +1,788 @@
+      JsonArray hits = searchResult
+        .getAsJsonObject("hits")
+        .getAsJsonArray("hits");
+      for (JsonElement hit:hits){
```
Weird spacing, please reformat one more time.

jkff commented on sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:
```diff
@@ -0,0 +1,403 @@
+      numDocsProcessed++;
+      // test every 100 docs to avoid overloading ES
+      if ((numDocsProcessed % 100) == 0) {
+        // there should be only one bundle because minimum BATCH_SIZE_MEGABYTES is 1MB,
```
As-is, the test is verifying only the case when `withMaxBatchSizeMegabytes` has no effect on batching because there are too few documents. E.g. the test would have passed if the code ignored this setting entirely.

Can you specify a smaller value of `withMaxBatchSizeMegabytes` and verify that it causes flushing when enough documents are accumulated?

jkff replied:
Changing unit to bytes seems reasonable to me.

echauchot commented:
Hi @jkff, thanks for the review. I addressed all your latest comments. PR is now ready for another (last I hope :) ) round of review. I have also fixed travis tests that behaved differently on some jdk.  @ssisk @jbonofre  please note guys, that I have moved IT material to an in preparation BEAM-1184-ELASTICSEARCH-IO PR.

echauchot commented:
@jkff since last good build, I refactored the unit tests to deduplicate code with ongoing integration tests (in preparation BEAM-1184-ELASTICSEARCH-IO PR.). I'll ping you when the build is ok again so that you can start another round of review

echauchot commented:
@jkff, everything is now ok. Can you please do the review? I Hope the code is near perfect now :) Thanks again!

jbonofre commented:
I'm starting a new review as well.

echauchot commented:
yes, thanks JB!

echauchot commented:
rebased on master

echauchot commented:
@jkff, I merged your PR, I also did the minor modifications that I mentioned in my review. I tested, everything is ok. I just need to deal with the post graduation changes and the IO will be ready to be merged.

jbonofre commented:
Rebasing and updating to have Jenkins and CI pass.

jbonofre commented:
Just rebased on master with post graduation renaming.

echauchot commented:
Thanks JB!
